File: test/patchbay_web/forum/relative_time_test.exs
defmodule PatchbayWeb.Forum.RelativeTimeTest do
  use ExUnit.Case, async: true

  alias PatchbayWeb.Forum.RelativeTime

  @now ~U[2026-09-01 12:00:00.000000Z]

  defp words_for(seconds), do: RelativeTime.in_words(DateTime.add(@now, -seconds), @now)

  test "a moment that never happened has no words" do
    assert RelativeTime.in_words(nil, @now) == "never"
  end

  test "anything inside the last minute reads as just now" do
    assert words_for(0) == "just now"
    assert words_for(59) == "just now"
  end

  test "a clock that runs ahead of the board still reads as just now" do
    assert RelativeTime.in_words(DateTime.add(@now, 90), @now) == "just now"
  end

  test "counts up through minutes, hours, days, weeks, months and years" do
    assert words_for(60) == "1 minute ago"
    assert words_for(59 * 60) == "59 minutes ago"
    assert words_for(3_600) == "1 hour ago"
    assert words_for(23 * 3_600) == "23 hours ago"
    assert words_for(86_400) == "1 day ago"
    assert words_for(6 * 86_400) == "6 days ago"
    assert words_for(7 * 86_400) == "1 week ago"
    assert words_for(29 * 86_400) == "4 weeks ago"
    assert words_for(30 * 86_400) == "1 month ago"
    assert words_for(200 * 86_400) == "6 months ago"
    assert words_for(365 * 86_400) == "1 year ago"
    assert words_for(3 * 365 * 86_400) == "3 years ago"
  end
end

File: platform/assets/test/hooks/relative_time.test.js
import assert from "node:assert/strict"
import {test} from "node:test"

import {inWords} from "../../js/hooks/relative_time.js"

const NOW = Date.parse("2026-09-01T12:00:00Z")

function ago(seconds) {
  return new Date(NOW - seconds * 1000).toISOString()
}

test("anything under a minute old reads as just now", () => {
  assert.equal(inWords(ago(0), NOW), "just now")
  assert.equal(inWords(ago(59), NOW), "just now")
})

test("a moment still in the future reads as just now", () => {
  assert.equal(inWords(ago(-10), NOW), "just now")
})

test("counts in the largest whole unit that fits", () => {
  assert.equal(inWords(ago(60), NOW), "1 minute ago")
  assert.equal(inWords(ago(180), NOW), "3 minutes ago")
  assert.equal(inWords(ago(3_600), NOW), "1 hour ago")
  assert.equal(inWords(ago(7_200), NOW), "2 hours ago")
  assert.equal(inWords(ago(86_400), NOW), "1 day ago")
  assert.equal(inWords(ago(604_800), NOW), "1 week ago")
  assert.equal(inWords(ago(2_592_000), NOW), "1 month ago")
  assert.equal(inWords(ago(31_536_000), NOW), "1 year ago")
})

test("says nothing at all when the timestamp is unreadable", () => {
  assert.equal(inWords("not a timestamp", NOW), "")
  assert.equal(inWords(null, NOW), "")
})
